### PR TITLE
rddCoerce and bindings to repartition, treeAggregate and binaryRecords

### PR DESCRIPTION
**Pull request by facundominguez**

### CODE REVIEW

File: src/Control/Distributed/Spark/RDD.hs [hunk 1]
```diff
@@ -112,6 +163,17 @@ textFile sc path = do
...
 
+-- recordLength = number of bytes
+binaryRecords :: SparkContext -> FilePath -> Int32 -> IO (RDD ByteString)
+binaryRecords sc fp recordLength = do
+  jpath <- reflect (Text.pack fp)
+  call sc "binaryRecords" [coerce jpath, coerce recordLength]
+
+-- | Reinterprets an @RDD a@ as an @RDD b@ provided that both types have
+-- the same representation on the java side.
+rddCoerce :: (Interp a ~ Interp b) => RDD a -> RDD b
```
Comment (mboes) [inline src/Control/Distributed/Spark/RDD.hs]: Do we really need this one? If we define this on RDD's then why not on any Java object newtype wrapper? Isn't this simply a case of calling `Data.Coerce.coerce` directly at the call site? Note that `RDD` is not abstract.

File: src/Control/Distributed/Spark/RDD.hs [hunk 2]
```diff
@@ -92,6 +122,27 @@ aggregate seqOp combOp zero rdd = do
   res :: JObject <- call rdd "aggregate" [coerce jzero, coerce jseqOp, coerce jcombOp]
   reify (unsafeCast res)
 
+treeAggregate :: ( Reflect (Closure (b -> a -> b)) ty1
```
Comment (mboes) [inline src/Control/Distributed/Spark/RDD.hs]: Let's keep the indentation consistent with what the rest of the module already uses.

File: src/Control/Distributed/Spark/RDD.hs (hunk 1 above)
Comment (facundominguez) [inline src/Control/Distributed/Spark/RDD.hs]: That could work. But doesn't `Data.Coerce.coerce` allow to coerce any pointer to a java object indiscriminately? This call would only allow to do that when the representations in java are the same.

Comment (mboes) [inline src/Control/Distributed/Spark/RDD.hs]: In that case there's a function missing in `Language.Java`. Better to factor this out into a common location than redefine it for each type out there.

**Comment by facundominguez** [comment]
The indentation change was squashed.

File: src/Control/Distributed/Spark/RDD.hs (hunk 1 above)
Comment (mboes) [inline src/Control/Distributed/Spark/RDD.hs]: `coerce1` or somesuch, by analogy with `Data.Typeable.gcast1`.

**Comment by mboes** [comment]
@facundominguez merged without the latest commit about rddCoerce. It's orthogonal to the other stuff anyways, so since it's attracting discussion of its own let's deal with that in a separate PR.

**Comment by facundominguez** [comment]
Works for me. How would the coercion look like in `Language.Java`?

**Pull request merged**